src/aleph/web/controllers/messages.py

```diff
@@ -237,50 +240,90 @@ async def _send_history_to_ws(
+) -> bool:
+    if message_type := query_params.message_type:
+        if message.type != message_type:
+            return False
+
+    # For simple filters, this reduces the amount of boilerplate
+    filters_by_message_field = {
+        "sender": "addresses",
+        "type": "message_type",
+        "item_hash": "hashes",
+        "ref": "refs",
+        "chain": "chains",
+        "channel": "channels",
+    }
+
+    for message_field, query_field in filters_by_message_field.items():
+        if user_filters := getattr(query_params, query_field):
+            if not isinstance(user_filters, list):
+                user_filters = [user_filters]
+            if not getattr(message, message_field) in user_filters:
+                return False
+
+    # Process filters on content and content.content
+    message_content = message.content
+    if content_types := query_params.content_types:
+        content_type = getattr(message_content, "type", None)
+        if content_type not in content_types:
+            return False
+
+    if content_hashes := query_params.content_hashes:
+        content_hash = getattr(message_content, "item_hash", None)
+        if content_hash not in content_hashes:
+            return False
+
+    # For tags, we only need to match one filter
+    if query_tags := query_params.tags:
+        nested_content = getattr(message.content, "content")
+        if not nested_content:
+            return False
+
+        content_tags = set(getattr(nested_content, "tags", []))
+        if (content_tags & set(query_tags)) == set():
+            return False
+
+    return True
+
+
 async def _start_mq_consumer(
     ws: aiohttp.web_ws.WebSocketResponse,
     mq_queue: aio_pika.abc.AbstractQueue,
-    session_factory: DbSessionFactory,
-    message_filters: Dict[str, Any],
+    query_params: WsMessageQueryParams,
 ) -> aio_pika.abc.ConsumerTag:
     """
     Starts the consumer task responsible for forwarding new aleph.im messages from
     the processing pipeline to a websocket.
 
     :param ws: Websocket.
     :param mq_queue: Message queue object.
-    :param session_factory: DB session factory.
-    :param message_filters: Filters to apply to select outgoing messages.
+    :param query_params: Message filters specified by the caller.
     """
 
     async def _process_message(mq_message: aio_pika.abc.AbstractMessage):
-        item_hash = aleph_json.loads(mq_message.body)["item_hash"]
-        # A bastardized way to apply the filters on the message as well.
-        # TODO: put the filter key/values in the RabbitMQ message?
-        with session_factory() as session:
-            matching_messages = get_matching_messages(
-                session=session,
-                hashes=[item_hash],
-                include_confirmations=True,
-                **message_filters,
-            )
+        payload_bytes = mq_message.body
+        payload_dict = aleph_json.loads(payload_bytes)
```

Comment:
Not related to this PR, but `aleph_json` looks like the project uses its own kind of JSON format.

Reply:
I know, but I wanted to make calls to the JSON module wrapper explicit. If you have a better idea, don't hesitate.